x/poolmanager/types/gov_test.go

```diff
@@ -0,0 +1,146 @@
+package types_test
+
+import (
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	proto "github.com/gogo/protobuf/proto"
+	"github.com/stretchr/testify/require"
+
+	"github.com/osmosis-labs/osmosis/v19/x/poolmanager/types"
+)
+
+func TestDenomPairTakerFeeProposalMarshalUnmarshal(t *testing.T) {
+	records := []types.DenomPairTakerFee{
+		{
+			Denom0:   "uion",
+			Denom1:   "uosmo",
+			TakerFee: sdk.MustNewDecFromStr("0.0013"),
+		},
+		{
+			Denom0:   "stake",
+			Denom1:   "uosmo",
+			TakerFee: sdk.MustNewDecFromStr("0.0016"),
+		},
+		{
+			Denom0:   "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
+			Denom1:   "uosmo",
+			TakerFee: sdk.MustNewDecFromStr("0.0017"),
+		},
+	}
+
+	tests := []struct {
+		proposal *types.DenomPairTakerFeeProposal
+	}{
+		{ // empty title
+			proposal: &types.DenomPairTakerFeeProposal{
+				Title:       "",
+				Description: "proposal to add denom pair taker fee records",
+			},
+		},
+		{ // empty description
+			proposal: &types.DenomPairTakerFeeProposal{
+				Title:       "title",
+				Description: "",
+			},
+		},
+		{ // happy path
+			proposal: &types.DenomPairTakerFeeProposal{
+				Title:             "title",
+				Description:       "proposal to add denom pair taker fee records",
+				DenomPairTakerFee: records,
+			},
+		},
+	}
+
+	for _, test := range tests {
+		bz, err := proto.Marshal(test.proposal)
+		require.NoError(t, err)
+		decoded := types.DenomPairTakerFeeProposal{}
+		err = proto.Unmarshal(bz, &decoded)
+		require.NoError(t, err)
+		require.Equal(t, *test.proposal, decoded)
+	}
+}
+
+func TestDenomPairTakerFeeProposal_ValidateBasic(t *testing.T) {
```

Comment:
I'd like to come back to this later, where I feel like we can make some more general method to get these guarantees